### DISCUSSION TITLE
Is it possible to have ComponentArray with differenty types?

### DISCUSSION THREAD

**Discussion by paolo-mgi** [jonniedie/ComponentArrays.jl#262] [Q&A] [answered]
I would like to integrate a system of differential equations in which the evolution of a real component depends upon an indicator of the remaining complex components.

`test_u0 = ComponentArray(v=1.0 + 0.0*im, 0.0+im*0.0], m= 1.0)` 
`function test_psi!(du, u, p, t)`
     `du.v = -im * u.v`
     `du.m =  u.m * real(dot(u.v,u.v))`
`end`
`test_prob = ODEProblem(test_psi!, test_u0, (0.0, tf), p)`
`test_sol=solve(test_prob, Vern7())`

Is it possible to use ComponentArray so that u.m is Float64 at every step? The code above returns

`ComponentVector{ComplexF64}(v = ComplexF64[0.9803381875004509 - 0.19732470228080312im, 0.0 + 0.0im], m = ComplexF64[1.2197284031519102 + 0.0im])'

I understand that ArrayPartition 

`psi0= [1.0 + 0.0im, 0.0 + 0.0im]`
`ArrayPartition(psi0,1.0)`

yields the desired output but I do not understand how to pass this output to an ODEProblem
Many thanks

**Comment by jonniedie** [comment]
You could technically do this by doing
```julia
test_u0 = ComponentArray{Union{Float64, ComplexF64}(v=[1.0 + 0.0*im, 0.0+im*0.0], m= 1.0)
```
But it’s probably going to be slightly slower than keeping everything complex and converting `m` to a real number when you unpack it. There’s unfortunately not much I can do on the ComponentArrays side to make heterogeneous arrays faster when the solvers need to blindly iterate over the elements. But keeping the `Union` to only two options should make things reasonably fast.

**Comment by paolo-mgi** [comment]
Please bear with me but I thik I do not understand the syntax correctly. If I enter

`test_u0 = ComponentArray{Union{ComplexF64,Float64}}(v=[1.0 + 0.0*im, 0.0+im*0.0], m= 1.0)`

I get 

`ComponentVector{Union{Float64, ComplexF64}}(v = Union{Float64, ComplexF64}[1.0 + 0.0im, 0.0 + 0.0im], m = 1.0 + 0.0im)`

which does not look to me the desired result. Many thanks!

**Reply by jonniedie** [reply, marked as answer]
Interesting. ~The solver must be doing something to promote the types then. I don’t think there’s much we can do about that.~ It looks like ComponentArrays is converting that type. But either way, is there any reason you can’t use a complex number for `m` and just convert it to real whenever you want to use it? That will be the most performant option.

**Comment by paolo-mgi** [comment]
I can definitely work with complex, as you suggest. I was just curious. ComponentArrays allowed me to finally get a quantum trajectory code to work properly :). I wanted only to further test its functionalyity. Many thanks for your work!
